### PR TITLE
[FIX] purchase: prevent changing product company after use in PO

### PR DESCRIPTION
**Pull request by jjmaksoud**
A product used in a PO line in a certain company should
not be allowed to change companies later on.

Description of the issue/feature this PR addresses:
The company restriction in PO lines / products is currently 1 way: a user cannot 
add products to a PO that belong to a different company, but they can update
the product later and create the same undesired company mismatch

Current behavior before PR:
To reproduce: 
- have a multi company environment with companies A and Bg 
- create a product in company A
- create a PO in company A and add the product
- edit the product to belong to company B
- no errors
- add another line to the PO using a valid product
- an error occurs 

plus it is causing upgrade failures

Desired behavior after PR is merged:
There will be no company mismatch between products and PO lines

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/purchase/models/product.py [hunk 1]
```diff
@@ -48,6 +49,14 @@ def action_view_po(self):
         }
         return action
 
+    @api.onchange('company_id')
+    def _onchange_company_id(self):
+        if self.company_id:
+            po_company_ids = self.env['purchase.order.line'].search([('product_id', 'in', self.product_variant_ids.ids), ('company_id', '!=', self.company_id.id)])
```
Comment (aj-fuentes) [inline addons/purchase/models/product.py:55]: Don't we need here the term `('company_id', '!=', False)`?

File: addons/purchase/models/product.py [hunk 2]
```diff
@@ -48,6 +49,14 @@ def action_view_po(self):
...
+            if po_company_ids:
+                raise ValidationError(_("You may not change the company on this product to %s because it "
+                                        "is already in use by Purchase Orders within this company:\n%s")
+                                        % (self.company_id.display_name, ', '.join(po_company_ids.mapped("order_id").mapped("name"))))
```
Comment (aj-fuentes) [inline addons/purchase/models/product.py:59]: The list may be long, add `limit=1` to the search above then
```suggestion
                                        "is already in use by the following Purchase Orders within this company:\n%s")
                                        % (self.company_id.display_name, ', '.join(po_company_ids.order_id.name))))
```

File: addons/purchase/models/product.py (hunk 1 above)
Comment (jjmaksoud) [reply addons/purchase/models/product.py:55]: company id is required for PO lines, should I add the `!= False` anyway?

File: addons/purchase/models/product.py [hunk 3]
```diff
@@ -48,6 +49,14 @@ def action_view_po(self):
         }
         return action
 
+    @api.onchange('company_id')
+    def _onchange_company_id(self):
+        if self.company_id:
+            po_company_ids = self.env['purchase.order.line'].search([('product_id', 'in', self.product_variant_ids.ids), ('company_id', '!=', self.company_id.id)], limit=1)
+            if po_company_ids:
+                raise ValidationError(_("You may not change the company on this product to %s because it "
+                                        "is already in use by the following Purchase Orders within this company:\n%s")
```
Comment (aj-fuentes) [inline addons/purchase/models/product.py:58]:
```suggestion
                                        "is already in use by the following Purchase Order within this company:\n%s")
```

**Review by aj-fuentes — APPROVED** [review]
LGTM, but I cannot decide if this is 100% correct, final approval is from @odoo/rd-logistics.

**Comment by Whenrow** [comment]
I would use _check_company on Purchase Orders instead https://www.odoo.com/documentation/15.0/developer/reference/backend/orm.html?highlight=check_company#odoo.models.BaseModel._check_company_auto

**Comment by jjmaksoud** [comment]
I tried using `check_company` on `product_id` fields, then `_check_company_auto` on the model, but the same behavior described in the issue persists --> The product id can be changed after the PO line is created

**Comment by Whenrow** [comment]
@jjmaksoud yes but If you leave the PO untouched what is the issue?

**Comment by jjmaksoud** [comment]
If the PO is edited -> error
If the PO is not edited, but the user has only the PO's company active and tries to view the PO -> error

**Pull request closed without merging**